File: test/ModularPipelines.UnitTests/Attributes/DotNetFormatOptionsTests.cs
using ModularPipelines.DotNet.Options;
using ModularPipelines.Helpers.Internal;

namespace ModularPipelines.UnitTests.Attributes;

public class DotNetFormatOptionsTests
{
    private readonly CommandModelProvider _modelProvider = new();
    private readonly CommandArgumentBuilder _argumentBuilder = new();

    private List<string> BuildArguments(object optionsObject)
    {
        var model = _modelProvider.GetCommandModel(optionsObject.GetType());
        return _argumentBuilder.BuildArguments(model, optionsObject);
    }

    [Test]
    public async Task ExcludeDiagnostics_Passes_Each_Id_Separately()
    {
        var options = new DotNetFormatOptions
        {
            ExcludeDiagnostics = ["CS0246", "CS1503"],
        };

        var args = BuildArguments(options);

        await Assert.That(args).IsEquivalentTo(new[]
        {
            "--exclude-diagnostics", "CS0246",
            "--exclude-diagnostics", "CS1503",
        });
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Attributes/DotNetFormatOptionsTests.fs
namespace ModularPipelines.UnitTests.FSharp.Attributes

open ModularPipelines.DotNet.Options
open ModularPipelines.Helpers.Internal
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

type DotNetFormatOptionsTests() =
    member private _.ModelProvider = CommandModelProvider()
    member private _.ArgumentBuilder = CommandArgumentBuilder()
    member private this.BuildArguments(optionsObject: obj) =
        let model = this.ModelProvider.GetCommandModel(optionsObject.GetType())
        this.ArgumentBuilder.BuildArguments(model, optionsObject)

    [<Test>]
    member this.ExcludeDiagnostics_Passes_Each_Id_Separately() = async {
        let options = DotNetFormatOptions()
        options.ExcludeDiagnostics <- [| "CS0246"; "CS1503" |]

        let args = this.BuildArguments(options)

        do! check(Assert.That((args |> Seq.toArray) = [|
            "--exclude-diagnostics"; "CS0246"
            "--exclude-diagnostics"; "CS1503"
        |]).IsTrue())
    }
